### PR TITLE
Forward to 3.0

### PR DESCRIPTION
**Pull request by Pierstoval**
Hey guys, now released :smile:

### CODE REVIEW

File: src/Symfony/Installer/NewCommand.php [hunk 1]
```diff
@@ -111,8 +111,8 @@ protected function checkSymfonyVersionIsInstallable()
         }
 
         // validate semver syntax
-        if (!preg_match('/^2\.\d(?:\.\d{1,2})?(?:-(?:dev|BETA\d*|RC\d*))?$/i', $this->version)) {
-            throw new \RuntimeException('The Symfony version must be 2.N or 2.N.M (where N and M are positive integers). The special "-dev", "-BETA" and "-RC" versions are also supported.');
+        if (!preg_match('/^[23]\.\d(?:\.\d{1,2})?(?:-(?:dev|BETA\d*|RC\d*))?$/i', $this->version)) {
+            throw new \RuntimeException('The Symfony version must be 2.N or 2.N.M or 3.N or 3.N.M (where N and M are positive integers). The special "-dev", "-BETA" and "-RC" versions are also supported.');
         }
 
         if (preg_match('/^2\.\d$/', $this->version)) {
```
Comment (stof) [inline src/Symfony/Installer/NewCommand.php]: this needs an update too

File: src/Symfony/Installer/NewCommand.php [hunk 2]
```diff
@@ -111,8 +111,8 @@ protected function checkSymfonyVersionIsInstallable()
         }
 
         // validate semver syntax
-        if (!preg_match('/^2\.\d(?:\.\d{1,2})?(?:-(?:dev|BETA\d*|RC\d*))?$/i', $this->version)) {
-            throw new \RuntimeException('The Symfony version must be 2.N or 2.N.M (where N and M are positive integers). The special "-dev", "-BETA" and "-RC" versions are also supported.');
+        if (!preg_match('/^[23]\.\d(?:\.\d{1,2})?(?:-(?:dev|BETA\d*|RC\d*))?$/i', $this->version)) {
+            throw new \RuntimeException('The Symfony version must be 2.N or 2.N.M or 3.N or 3.N.M (where N and M are positive integers). The special "-dev", "-BETA" and "-RC" versions are also supported.');
```
Comment (javiereguiluz) [inline src/Symfony/Installer/NewCommand.php]: I'm not sure but maybe rewording the error message as follows would make it more readable:

```
[...] must be 2.N, 2.N.M, 3.N or 3.N.M [...]
```

File: src/Symfony/Installer/NewCommand.php (hunk 1 above)
Comment (Pierstoval) [inline src/Symfony/Installer/NewCommand.php]: Fixed!

File: src/Symfony/Installer/NewCommand.php (hunk 2 above)
Comment (Pierstoval) [inline src/Symfony/Installer/NewCommand.php]: Changed for your proposal :)

**Comment by javiereguiluz** [comment]
@Pierstoval thanks for this! A new test for 3.0 version would be great. You just need to add a new item to this data provider: https://github.com/symfony/symfony-installer/blob/master/tests/Symfony/Installer/Tests/IntegrationTest.php#L96

**Comment by Pierstoval** [comment]
@javiereguiluz I added the test but it'd need more integration to make it succeed, because of Symfony3 different structure, so it'll need more updates on the tests :)

**Comment by javiereguiluz** [comment]
@Pierstoval and *that's* why the installer didn't have support for Symfony 3 yet. It's not a matter of changing just a number :)

**Comment by Pierstoval** [comment]
I'm on it ;)

**Comment by Pierstoval** [comment]
Done it!
 
Just, there's still a problem: the `composer.lock` stores a version code as `dev-master`, whereas it should be `3.0.0-BETA1` as well as any other beta version.
 
Plus, I don't know when it'll be available (I don't know this info abut the release process) but the http://symfony.com/versions.json should be updated in order to show the state of the `3.0` branch, but I don't know if we need a proper release for this or if the beta is sufficient.

**Comment by Pierstoval** [comment]
If needed, I can patch a workaround for the `composer.lock` version problem, but I don't know Symfony policy about 3.0 releases.
Anyway it would make the tests succeed and a new release of the installer could be done.

File: src/Symfony/Installer/DownloadCommand.php [hunk 3]
```diff
@@ -416,6 +431,16 @@ protected function isEmptyDirectory($dir)
...
     }
 
+    /**
+     * Checks that the asked version is in the 3.x branch.
+     *
+     * @return bool
+     */
+    protected function isSymfony3()
+    {
+        return strpos($this->version, '3') === 0;
```
Comment (xabbuh) [inline src/Symfony/Installer/DownloadCommand.php]:
```php
return '3' === $this->version[0];
```

File: src/Symfony/Installer/NewCommand.php [hunk 4]
```diff
@@ -260,10 +262,12 @@ protected function displayInstallationResult()
             ));
         }
 
+        $consoleDir = ($this->isSymfony3() ? 'bin' : 'app');
+
         $this->output->writeln(sprintf(
             "    * Configure your application in <comment>app/config/parameters.yml</comment> file.\n\n".
             "    * Run your application:\n".
-            "        1. Execute the <comment>php app/console server:run</comment> command.\n".
+            "        1. Execute the <comment>php $consoleDir/console server:run</comment> command.\n".
```
Comment (xabbuh) [inline src/Symfony/Installer/NewCommand.php]: imo you should use a `%s` placeholder here

Comment (Pierstoval) [inline src/Symfony/Installer/NewCommand.php]: Done, thanks!

File: src/Symfony/Installer/DownloadCommand.php (hunk 3 above)
Comment (Pierstoval) [inline src/Symfony/Installer/DownloadCommand.php]: Fixed, thanks! I didn't know this syntax would be fully compatible :)

**Comment by javiereguiluz** [comment]
I've tested this PR and everything works fine for me. I'll merge it soon unless someone raises a valid concern.

@Pierstoval I'd like to do just a minor change:

```diff
-                "    * Change your current directory to <comment>%s</comment>\n\n", $this->projectDir
+                "    * Change your current directory to <comment>%s</comment>\n", $this->projectDir
```

This change prevents the ugly double blank line in the command output:

excessive_blank_lines

**Comment by Pierstoval** [comment]
Done in eba3537 :smile:

**Comment by javiereguiluz** [comment]
@Pierstoval thanks for working on this. I've merged the changes and I'll release a new version of the installer soon, so we can test everything well and it's ready for the big Symfony 2.8 and 3.0 launch next week. Thanks!
